cmd/embedded-cluster/restore.go

```diff
@@ -0,0 +1,585 @@
+// If a file already sits there, this function returns an error.
+func ensureK0sConfigForRestore(c *cli.Context) error {
+	cfgpath := defaults.PathToK0sConfig()
+	if _, err := os.Stat(cfgpath); err == nil {
+		return fmt.Errorf("configuration file already exists")
+	}
+	if err := os.MkdirAll(filepath.Dir(cfgpath), 0755); err != nil {
+		return fmt.Errorf("unable to create directory: %w", err)
+	}
+	cfg := config.RenderK0sConfig()
+	opts := []addons.Option{}
+	if err := config.UpdateHelmConfigsForRestore(cfg, opts...); err != nil {
+		return fmt.Errorf("unable to update helm configs: %w", err)
+	}
+	var err error
+	if cfg, err = applyUnsupportedOverrides(c, cfg); err != nil {
+		return fmt.Errorf("unable to apply unsupported overrides: %w", err)
+	}
+	data, err := k8syaml.Marshal(cfg)
+	if err != nil {
+		return fmt.Errorf("unable to marshal config: %w", err)
+	}
+	fp, err := os.OpenFile(cfgpath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return fmt.Errorf("unable to create config file: %w", err)
+	}
+	defer fp.Close()
+	if _, err := fp.Write(data); err != nil {
+		return fmt.Errorf("unable to write config file: %w", err)
+	}
+	return nil
+}
+
+// runOutroForRestore calls Outro() in all enabled addons for restore operations by means of Applier.
+func runOutroForRestore(c *cli.Context) error {
+	os.Setenv("KUBECONFIG", defaults.PathToKubeConfig())
+	opts := []addons.Option{}
+	return addons.NewApplier(opts...).OutroForRestore(c.Context)
+}
+
+func isBackupRestorable(backup *velerov1.Backup, rel *release.ChannelRelease) (bool, string) {
+	if backup.Annotations["kots.io/embedded-cluster"] != "true" {
+		return false, "is not an embedded cluster backup"
+	}
+	if v := backup.Annotations["kots.io/embedded-cluster-version"]; v != defaults.Version {
+		return false, fmt.Sprintf("has a different embedded cluster version (%q) than the current version (%q)", v, defaults.Version)
+	}
+	if backup.Status.Phase != velerov1.BackupPhaseCompleted {
+		return false, fmt.Sprintf("has a status of %q", backup.Status.Phase)
+	}
+	if _, ok := backup.Annotations["kots.io/apps-versions"]; !ok {
+		return false, "is missing the kots.io/apps-versions annotation"
+	}
+	appsVersions := map[string]string{}
+	if err := json.Unmarshal([]byte(backup.Annotations["kots.io/apps-versions"]), &appsVersions); err != nil {
+		return false, "unable to json parse kots.io/apps-versions annotation"
+	}
+	if len(appsVersions) == 0 {
+		return false, "has no applications"
+	}
+	if len(appsVersions) > 1 {
+		return false, "has more than one application"
+	}
+	if _, ok := appsVersions[rel.AppSlug]; !ok {
+		return false, fmt.Sprintf("does not contain the %q application", rel.AppSlug)
+	}
+	if versionLabel := appsVersions[rel.AppSlug]; versionLabel != rel.VersionLabel {
+		return false, fmt.Sprintf("has a different app version (%q) than the current version (%q)", versionLabel, rel.VersionLabel)
+	}
+	return true, ""
```

Comment:
we'll need to add a 'get the correct version with this curl command' message here (and earlier) at some point in the future
not a blocker

Reply:
Good suggestion.